plugins/launcher_service_plugin/launcher_service_plugin.cpp

```diff
@@ -0,0 +1,1072 @@
+            FC_ASSERT(false, "Unknown action ${action} in contract ${contract}", ("action", action)( "contract", account ));
+         }
+
+         char temp[512 * 1024];
+         fc::datastream<char*> ds(temp, sizeof(temp)-1 );
+         abis->variant_to_binary( action_type, action_args_var, ds, _config.abi_serializer_max_time );
+         std::vector<char> r;
+         r.assign(temp, temp + ds.tellp());
+         return r;
+      }
+
+      fc::variant push_transaction(int cluster_id, int node_id, signed_transaction& trx,
+                                   const std::vector<public_key_type> &sign_keys = std::vector<public_key_type>(),
+                                   packed_transaction::compression_type compression = packed_transaction::compression_type::none) {
+         int port = _running_clusters[cluster_id].nodes[node_id].http_port;
+         auto &info = _running_clusters[cluster_id].nodes[node_id].get_info_cache;
+         if (fc::time_point::now() >= _running_clusters[cluster_id].nodes[node_id].get_info_cache_time + fc::milliseconds(500)) {
+            fc::variant info_ = get_info(cluster_id, node_id);
+            info = info_.as<get_info_results>();
+            _running_clusters[cluster_id].nodes[node_id].get_info_cache_time = fc::time_point::now();
+         }
+
+         if (trx.signatures.size() == 0) {
+            trx.expiration = info.head_block_time + fc::seconds(3);
+
+            // Set tapos, default to last irreversible block if it's not specified by the user
+            block_id_type ref_block_id = info.last_irreversible_block_id;
+            trx.set_reference_block(ref_block_id);
+
+           // if (tx_force_unique) {
+           //    trx.context_free_actions.emplace_back( generate_nonce_action() );
+           // }
+
+            trx.max_cpu_usage_ms = 30;
+            trx.max_net_usage_words = 50000;
+            trx.delay_sec = 0;
+         }
+
+         bool has_key = false;
+         if (sign_keys.size()) {
+            for (const public_key_type &pub_key : sign_keys) {
+               if (_running_clusters[cluster_id].imported_keys.find(pub_key) != _running_clusters[cluster_id].imported_keys.end()) {
+                  private_key_type pri_key = _running_clusters[cluster_id].imported_keys[pub_key];
+                  trx.sign(pri_key, *(chain_id_type *)&(info.chain_id));
+                  has_key = true;
+               } else {
+                  throw std::runtime_error("private key of \"" + (std::string)pub_key + "\" not imported");
+               }
+            }
+         } else {
+            fc::variant required_keys = determine_required_keys(cluster_id, node_id, trx);
+            const fc::variant &keys = required_keys["required_keys"];
+            for (const fc::variant &k : keys.get_array()) {
+               public_key_type pub_key = public_key_type(k.as_string());
+               private_key_type pri_key = _running_clusters[cluster_id].imported_keys[pub_key];
```

Comment:
This would create an entry if `pub_key` does not exist in the map. Instead use the `find` pattern as is done above. If a required key cannot be found then throw an exception as above. Then there will be no need for the `has_key` boolean at all.

Reply:
should be fixed.

Reply:
It doesn't appear to be fixed yet.

Reply:
ah, my bad. please check it again.